cmd/panfs-multipart.go

```diff
@@ -872,33 +1003,68 @@ func (fs *PANFSObjects) AbortMultipartUpload(ctx context.Context, bucket, object
 		return toObjectErr(err, bucket)
 	}
 
+	handleStatMetaFileError := func(e error) error {
+		// caller is responsible to check that e is not nil
+		if err == errFileNotFound || err == errFileAccessDenied {
+			return InvalidUploadID{Bucket: bucket, Object: object, UploadID: uploadID}
+		}
+		return toObjectErr(err, bucket, object)
+	}
+
+	uploadIDDir := fs.getUploadIDDir(bucketPath, object, uploadID)
+	_, err = fsStatFile(ctx, pathJoin(uploadIDDir, fs.metaJSONFile))
+	if err != nil {
+		return handleStatMetaFileError(err)
+	}
+
 	fs.appendFileMapMu.Lock()
 	// Remove file in tmp folder
 	file := fs.appendFileMap[uploadID]
-	if file != nil {
-		fsRemoveFile(ctx, file.filePath)
+	if file == nil {
+		file = fs.getPanFSMultipartAppendFile(bucketPath, object, uploadID)
+		fs.appendFileMap[uploadID] = file
 	}
-	delete(fs.appendFileMap, uploadID)
 	fs.appendFileMapMu.Unlock()
 
-	uploadIDDir := fs.getUploadIDDir(bucketPath, bucket, object, uploadID)
-	// Just check if the uploadID exists to avoid copy if it doesn't.
+	if err = file.flock.Lock(); err != nil {
+		return InvalidUploadID{Bucket: bucket, Object: object, UploadID: uploadID}
+	}
+	defer func() {
+		if err == nil {
+			file.flock.Unlock()
+			// remove lock file
+			fsRemoveFile(ctx, fs.getMultipartLockFile(bucketPath, object, uploadID))
+			// Delete append file from memory only if no error occurred during mkdir and rename on next steps
+			fs.appendFileMapMu.Lock()
+			delete(fs.appendFileMap, uploadID)
+			fs.appendFileMapMu.Unlock()
+		} else {
+			file.flock.Unlock()
+		}
+	}()
+
+	// ensure that uploadId dir is still here
 	_, err = fsStatFile(ctx, pathJoin(uploadIDDir, fs.metaJSONFile))
 	if err != nil {
-		if err == errFileNotFound || err == errFileAccessDenied {
-			return InvalidUploadID{Bucket: bucket, Object: object, UploadID: uploadID}
-		}
-		return toObjectErr(err, bucket, object)
+		return handleStatMetaFileError(err)
 	}
+	fsRemoveFile(ctx, file.filePath)
```

Comment:
Please, remove the `mpart.json` file first.

Reply:
That file will be removed on the next lines:
```
defer fsRemoveAll(ctx, fsTmpObjPath)

	err = Rename(uploadIDDir, fsTmpObjPath)
```
Do you think it requires to delete it manually? This file is only accessed under the flock so no one will read it during aborting process